File: swift/Sources/Aksara/TranslationTable.swift
import Foundation

/// An **immutable** snapshot of one language's translations: the flattened
/// dot-path → template map plus optional metadata.
///
/// The whole OTA/live-update design hinges on immutability — updates build a brand
/// new `TranslationTable` and swap a single reference. Nothing is ever mutated in
/// place, so a `t(...)` in flight always sees a consistent table.
public struct TranslationTable: Sendable {
    public let language: String
    /// Flattened `dot.path` → raw template (with `{{var}}` placeholders intact).
    public let entries: [String: String]
    /// Optional `_version` from the bundle, used as a cheap change signal alongside ETag.
    public let version: String?

    public init(language: String, entries: [String: String], version: String? = nil) {
        self.language = language
        self.entries = entries
        self.version = version
    }

    public var count: Int { entries.count }

    @inline(__always)
    public func value(for key: String) -> String? { entries[key] }
}

extension TranslationTable {
    /// Validate + parse `data` into an immutable table. Throws `SchemaValidationError`
    /// if the payload isn't a well-formed translation bundle.
    static func make(language: String, data: Data) throws -> TranslationTable {
        var json = try SchemaValidator.validated(data)

        // Pull `_version` out of the graph so it doesn't pollute the key space.
        var version: String?
        if let raw = json.removeValue(forKey: "_version") {
            if let s = raw as? String { version = s }
            else if let n = raw as? NSNumber { version = n.stringValue }
        }

        let entries = Flattener.flatten(json)
        return TranslationTable(language: language, entries: entries, version: version)
    }
}

File: kotlin/aksara-core/src/main/kotlin/com/aksara/TranslationTable.kt
package com.aksara

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * An **immutable** snapshot of one language's translations: the flattened dot-path
 * -> template map plus optional metadata.
 *
 * The whole OTA/live-update design hinges on immutability — updates build a brand new
 * table and swap a single reference. Nothing is mutated in place, so a `t(...)` in
 * flight always sees a consistent table.
 *
 * Mirror of the Swift `TranslationTable`.
 */
data class TranslationTable(
    val language: String,
    val entries: Map<String, String>,
    val version: String? = null,
) {
    val count: Int get() = entries.size

    fun value(key: String): String? = entries[key]

    companion object {
        /**
         * Validate + parse [bytes] into an immutable table. Throws
         * [SchemaValidationException] if the payload isn't a well-formed bundle.
         */
        fun make(language: String, bytes: ByteArray): TranslationTable {
            val json = SchemaValidator.validated(bytes)

            // Pull `_version` out of the graph so it doesn't pollute the key space.
            val mutable = json.toMutableMap()
            val version = (mutable.remove("_version") as? JsonPrimitive)?.content

            val entries = Flattener.flatten(JsonObject(mutable))
            return TranslationTable(language, entries, version)
        }
    }
}
